api/src/test/java/com/epam/pipeline/acl/pipeline/PipelineApiServiceTest.java

```diff
@@ -0,0 +1,831 @@
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER)
+    public void shouldNotLoadAllPipelinesWithoutVersionWhenPermissionIsNotGranted() {
+        initAclEntity(pipeline);
+        doReturn(mutableListOf(pipeline)).when(mockPipelineManager).loadAllPipelines(false);
+
+        assertThat(pipelineApiService.loadAllPipelinesWithoutVersion()).isEmpty();
+    }
+
+    @Test
+    @WithMockUser(roles = ADMIN_ROLE)
+    public void shouldDeletePipelineForAdmin() {
+        doReturn(pipeline).when(mockPipelineManager).delete(ID, true);
+
+        assertThat(pipelineApiService.delete(ID, true)).isEqualTo(pipeline);
+    }
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER, roles = PIPELINE_MANAGER_ROLE)
+    public void shouldDeletePipelineWhenPermissionIsGranted() {
+        initAclEntity(pipeline, AclPermission.WRITE);
+        doReturn(pipeline).when(mockPipelineManager).delete(ID, true);
+
+        assertThat(pipelineApiService.delete(ID, true)).isEqualTo(pipeline);
+    }
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER, roles = PIPELINE_MANAGER_ROLE)
+    public void shouldDenyDeletePipelineWhenPermissionIsNotGranted() {
+        initAclEntity(pipeline);
+        doReturn(pipeline).when(mockPipelineManager).delete(ID, true);
+
+        assertThrows(AccessDeniedException.class, () -> pipelineApiService.delete(ID, true));
+    }
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER)
+    public void shouldDenyDeletePipelineForUserRole() {
+        initAclEntity(pipeline, AclPermission.WRITE);
+        doReturn(pipeline).when(mockPipelineManager).delete(ID, true);
+
+        assertThrows(AccessDeniedException.class, () -> pipelineApiService.delete(ID, true));
+    }
+
+    @Test
+    @WithMockUser(roles = ADMIN_ROLE)
+    public void shouldLoadAllRunsByPipelineForAdmin() {
+        doReturn(pipelineRunList).when(mockPipelineRunManager).loadAllRunsByPipeline(ID);
+
+        assertThat(pipelineApiService.loadAllRunsByPipeline(ID)).isEqualTo(pipelineRunList);
+    }
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER)
+    public void shouldLoadAllRunsByPipelineWhenPermissionIsGranted() {
+        initAclEntity(pipeline, AclPermission.READ);
+        initAclEntity(pipelineRun, AclPermission.READ);
+        doReturn(pipelineRunList).when(mockPipelineRunManager).loadAllRunsByPipeline(ID);
+
+        final List<PipelineRun> returnedPipelineRunList = pipelineApiService.loadAllRunsByPipeline(ID);
+
+        assertThat(returnedPipelineRunList).hasSize(1).contains(pipelineRun);
+        assertThat(returnedPipelineRunList.get(0).getMask()).isEqualTo(READ_PERMISSION);
+    }
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER)
+    public void shouldDenyLoadAllRunsByPipelineWhenPermissionIsNotGranted() {
+        initAclEntity(pipeline);
+        doReturn(pipelineRunList).when(mockPipelineRunManager).loadAllRunsByPipeline(ID);
+
+        assertThrows(AccessDeniedException.class, () -> pipelineApiService.loadAllRunsByPipeline(ID));
+    }
+
```

Comment:
It looks like `PipelineApiService.loadAllRunsByPipeline` method may return runs for which a user doesn't have sufficient permissions.

Could you please check if this is really what's happening. If it is then we should add a test for it and create an issue asking for a fix.

Reply:
The problem appears when a user has permissions for a pipeline but doesn't have permissions for a run of that pipeline.